### PR TITLE
Allow to resolve `authenticationOptions` based on authenticator name or display name

### PR DESCRIPTION
**Pull request by ThaminduDilshan**
Currently in adaptive scripts we use authenticator friendly names to resolve `authenticationOptions` parameter. However this raises troubles when the authenticator display names get changed.
We should use authenticator names instead of friendly names in adaptive scripts. Friendly names should be expect to change.
To preserve backward compatibility, we are introducing a new parameter to resolve the authenticator identifier.

Example:
```js
executeStep(1, {
    authenticationOptions: [
        { idp: "LOCAL" }, 
        { authenticator: "sms-otp-authenticator" }, 
        { authenticatorNameResolver: "AUTHENTICATOR_NAME" }
    ]
}, {
    onSuccess: function (context) {
        // Do something on success
    }
});
```

### Related issue
- https://github.com/wso2/product-is/issues/14843

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java [hunk 1]
```diff
@@ -245,6 +245,7 @@ private void handleStepOptions(StepConfig stepConfig, Map<String, String> stepOp
     protected void filterOptions(Map<String, Map<String, String>> authenticationOptions, StepConfig stepConfig) {
 
         Map<String, Set<String>> filteredOptions = new HashMap<>();
+        final String[] authenticatorNameResolver = new String[1];
```
Comment (ThaminduDilshan) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java:248]: A string array is used here as non-final variables are not allowed to change inside lambda functions.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java [hunk 2]
```diff
@@ -280,19 +286,31 @@ protected void filterOptions(Map<String, Map<String, String>> authenticationOpti
...
+                                    removeOption = false;
+                                    break;
+                                }
+                            } else {
+                                // If the authenticatorNameResolver is not defined or is equal to the authenticator
+                                // display name, authenticator display name will be used to resolve the
+                                // authentication options.
+                                // Should translate the display name to actual name, and keep/remove option.
+                                if (authenticatorConfig.getName().equals(localAuthenticatorConfig.getName()) &&
+                                        authenticators.contains(localAuthenticatorConfig.getDisplayName())) {
```
Comment (ThaminduDilshan) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java:310]: `toLowerCase()` is removed here in favor of autheticator name.
i.e. `filteredOptions.get(idp).add(authenticator.toLowerCase());` need to remove authenticator being lower case here in favor of authenticator name.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java [hunk 3]
```diff
@@ -255,9 +256,14 @@ protected void filterOptions(Map<String, Map<String, String>> authenticationOpti
...
                 if (StringUtils.isNotBlank(authenticator)) {
-                    filteredOptions.get(idp).add(authenticator.toLowerCase());
+                    filteredOptions.get(idp).add(authenticator);
                 }
             }
+            if (StringUtils.isNotBlank(
+                    option.get(FrameworkConstants.JSAttributes.AUTHENTICATOR_NAME_RESOLVER_PARAM))) {
+                authenticatorNameResolver[0] = option.get(FrameworkConstants.JSAttributes.
+                        AUTHENTICATOR_NAME_RESOLVER_PARAM);
+            }
```
Comment (dewniMW) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java:266]:
```suggestion
            String authenticatorNameResolverParam = 
                    option.get(FrameworkConstants.JSAttributes.AUTHENTICATOR_NAME_RESOLVER_PARAM);
            if (StringUtils.isNotBlank(authenticatorNameResolverParam)) {
                authenticatorNameResolver[0] = authenticatorNameResolverParam;
            }
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java (hunk 2 above)
Comment (dewniMW) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java:310]: wouldn't this affect existing scripts if the authenticator display name used is case insensitive

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java [hunk 4]
```diff
@@ -280,19 +286,31 @@ protected void filterOptions(Map<String, Map<String, String>> authenticationOpti
...
+                            } else {
+                                // If the authenticatorNameResolver is not defined or is equal to the authenticator
+                                // display name, authenticator display name will be used to resolve the
+                                // authentication options.
+                                // Should translate the display name to actual name, and keep/remove option.
+                                if (authenticatorConfig.getName().equals(localAuthenticatorConfig.getName()) &&
+                                        authenticators.contains(localAuthenticatorConfig.getDisplayName())) {
+                                    removeOption = false;
+                                    break;
+                                }
```
Comment (dewniMW) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java:313]: what happens if the authenticatorNameResolver[0] contains an incorrect value such as "dummy" value?

Comment (ThaminduDilshan) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java:313]: It falls to the default path of using authenticator display name

Comment (dewniMW) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java:313]: in such cases, shall we add a debug log to indicate that the authenticatorNameResolver is invalid, hence proceeding with the authenticator display name

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java (hunk 2 above)
Comment (ThaminduDilshan) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java]: Fixed by first resolving the `authenticatorNameResolver` prior to looping through list to check authenticators

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java (hunk 3 above)
Comment (ThaminduDilshan) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java]: Fixed

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java (hunk 4 above)
Comment (ThaminduDilshan) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java:336]: Fixed

**Comment by ThaminduDilshan** [comment]
Change of approach

**Pull request closed without merging**